girder_worker_utils/decorators.py

```diff
@@ -16,13 +17,190 @@ class MissingInputException(Exception):
+
+    def _get_class(self, p):
+        if self._is_varargs(p):
+            return Varargs
+        elif self._is_kwargs(p):
+            return Kwargs
+        elif self._is_posarg(p):
+            return Arg
+        elif self._is_kwarg(p):
+            return KWArg
+        else:
+            raise RuntimeError("Could not determine parameter type!")
+
+
+    def set_metadata(self, name, key, value):
+        if name not in self._signature.parameters:
+            raise RuntimeError("{} is not a valid argument to this function!")
+
+        if name not in self._metadata:
+            self._metadata[name] = {}
+
+        self._metadata[name][key] = value
+
+    @property
+    def arguments(self):
+        # Only return arguments if we've declared them as paramaters
+        # This prevents us from returning things like 'self' of bound
+        # methods (e.g. celery tasks) etc.  This is a dubious design
+        # decision.
+        return [
+            self._construct_argument(
+                self._get_class(self._signature.parameters[name]), name)
+            for name in self._signature.parameters if name in self._metadata]
+
+    @property
+    def varargs(self):
+        for name in self._signature.parameters:
+            if name in self._metadata and \
+               self._is_varargs(self._signature.parameters[name]):
+                return self._construct_argument(Varargs, name)
+        return None
+
+    @property
+    def kwargs(self):
+        for name in self._signature.parameters:
+            if name in self._metadata and \
+               self._is_kwargs(self._signature.parameters[name]):
+                return self._construct_argument(Kwargs, name)
+        return None
+
+    @property
+    def positional_args(self):
+        return [arg for arg in self.arguments if isinstance(arg, Arg)]
+
+    @property
+    def keyword_args(self):
+        return [arg for arg in self.arguments if isinstance(arg, KWArg)]
+
+
+def parameter(name, **kwargs):
+    if not isinstance(name, six.string_types):
+        raise TypeError('Expected argument name to be a string')
+
+    data_type = kwargs.get("data_type", None)
+    if data_type is not None and callable(data_type):
+        kwargs['data_type'] = data_type(name, **kwargs)
+
+    def argument_wrapper(func):
+        if not hasattr(func, GWFuncDesc._func_desc_attr):
+            setattr(func, GWFuncDesc._func_desc_attr, GWFuncDesc(func))
+
+        desc = getattr(func, GWFuncDesc._func_desc_attr)
+
+        # Make sure the metadata key exists even if we don't set any
+        # values on it.  This ensures that metadata's keys represent
+        # the full list of parameters that have been identified by the
+        # user (even if there is no actual metadata associated with
+        # the argument).
+        desc._metadata[name] = {}
+
+        for key, value in six.iteritems(kwargs):
+            desc.set_metadata(name, key, value)
```

Comment:
It seems weird to directly set `_metadata` and then use `set_metadata` to set its values.

Reply:
It may be that there are no key/value pairs,  i can add an init_metadata method to be more clear